File: tests/java/Nested.java
package com.example.nested;

public class Outer {

    public static final int VERSION = 3;

    private int outerField;

    public Outer(int outerField) {
        this.outerField = outerField;
    }

    public int getOuterField() {
        return outerField;
    }

    public static class StaticNested {

        private String name;

        public StaticNested(String name) {
            this.name = name;
        }

        public String getName() {
            return name;
        }

        public class DeepInner {

            public void greet() {
                System.out.println("Hello from " + name);
            }
        }
    }

    public class Inner {

        public static final String LABEL = "inner";

        private int value;

        public Inner(int value) {
            this.value = value;
        }

        public int combined() {
            return outerField + value;
        }
    }

    public interface Callback {
        void onEvent(String event);
        default void onError(String msg) {
            System.err.println(msg);
        }
    }

    public enum Priority {
        LOW,
        MEDIUM,
        HIGH;

        public boolean isUrgent() {
            return this == HIGH;
        }
    }

    public static Outer create(int v) {
        return new Outer(v);
    }
}

interface TopLevel {
    void execute();
}

File: tests/kotlin/nested.kt
package com.example.nested

const val ROOT_VERSION = 1

class Outer(val name: String) {

    val outerProp: Int = 42

    fun outerMethod(): String {
        return name
    }

    inner class Inner(val value: Double) {
        fun innerMethod(): Double {
            return value * outerProp
        }

        object InnerSingleton {
            const val MAGIC = "inner-magic"
            fun create(): InnerSingleton = InnerSingleton
        }
    }

    class StaticNested {
        val count: Int = 0

        fun increment(): Int {
            return count + 1
        }

        class DeepNested {
            fun deepMethod() {}

            interface DeepInterface {
                fun deepAbstract(): Boolean
            }
        }
    }

    companion object {
        const val TAG = "Outer"
        var instanceCount: Int = 0

        fun newInstance(name: String): Outer {
            instanceCount++
            return Outer(name)
        }
    }

    interface Callback {
        fun onSuccess(result: String)
        fun onError(error: Throwable)
        fun onComplete()
    }

    enum class State {
        IDLE,
        LOADING,
        SUCCESS,
        ERROR,
    }

    sealed class Event {
        data class Success(val data: String) : Event()
        data class Failure(val error: Throwable) : Event()
        object Loading : Event()
    }
}

object TopLevelSingleton {
    val value: String = "singleton"

    fun getValue(): String {
        return value
    }

    object Nested {
        fun nestedFun() {}
    }
}
